src/particles.zig

```diff
@@ -0,0 +1,385 @@
+		defer main.stackAllocator.free(path);
+		textureIDs.append(arenaForWorld.allocator().dupe(u8, path));
+		readTextureData(path, &particleType);
+
+		particleTypeHashmap.put(id, @intCast(types.items.len)) catch unreachable;
+		types.append(particleType);
+	}
+
+	fn readTextureData(_path: []const u8, typ: *ParticleType) void {
+		const animationFrames = typ.animationFrames;
+		const path = _path[0 .. _path.len - ".png".len];
+		typ.startFrame = @intCast(textures.items.len);
+		const base = readTextureFile(path, ".png", Image.defaultImage);
+		const emission = readTextureFile(path, "_emission.png", Image.emptyImage);
+		for(0..animationFrames) |i| {
+			textures.append(extractAnimationSlice(base, i, animationFrames));
+			emissionTextures.append(extractAnimationSlice(emission, i, animationFrames));
+		}
+	}
+
+	fn extendedPath(_allocator: main.heap.NeverFailingAllocator, path: []const u8, ending: []const u8) []const u8 {
+		return std.fmt.allocPrint(_allocator.allocator, "{s}{s}", .{path, ending}) catch unreachable;
+	}
+
+	fn readTextureFile(_path: []const u8, ending: []const u8, default: Image) Image {
+		const path = extendedPath(main.stackAllocator, _path, ending);
+		defer main.stackAllocator.free(path);
+		return Image.readFromFile(arenaForWorld.allocator(), path) catch default;
+	}
+
+	fn extractAnimationSlice(image: Image, frame: usize, frames: usize) Image {
+		if(image.height < frames) return image;
+		var startHeight = image.height/frames*frame;
+		if(image.height%frames > frame) startHeight += frame else startHeight += image.height%frames;
+		var endHeight = image.height/frames*(frame + 1);
+		if(image.height%frames > frame + 1) endHeight += frame + 1 else endHeight += image.height%frames;
+		var result = image;
+		result.height = @intCast(endHeight - startHeight);
+		result.imageData = result.imageData[startHeight*image.width .. endHeight*image.width];
+		return result;
+	}
+
+	pub fn generateTextureArray() void {
+		std.log.debug("Particle texture sizes: {d} {d}", .{textures.items.len, emissionTextures.items.len});
+		textureArray.generate(textures.items, true, true);
+		c.glTexParameterf(c.GL_TEXTURE_2D_ARRAY, c.GL_TEXTURE_MAX_ANISOTROPY, @floatFromInt(main.settings.anisotropicFiltering));
+		emissionTextureArray.generate(emissionTextures.items, true, false);
+		c.glTexParameterf(c.GL_TEXTURE_2D_ARRAY, c.GL_TEXTURE_MAX_ANISOTROPY, @floatFromInt(main.settings.anisotropicFiltering));
+
+		particleTypesSSBO = SSBO.initStatic(ParticleType, ParticleManager.types.items);
+		particleTypesSSBO.?.bind(14);
+	}
+
+	pub fn update(deltaTime: f64) void {
+		const dt: f32 = @floatCast(deltaTime);
+		ParticleSystem.update(dt);
+	}
+
+	pub fn render(playerPosition: Vec3d, ambientLight: Vec3f) void {
+		ParticleSystem.render(playerPosition, ambientLight);
+	}
+};
+
+pub const ParticleSystem = struct {
+	pub const maxCapacity: u32 = 65536;
+	var particles: []Particle = undefined;
```

Comment:
We don't need to heap-allocate this
```suggestion
	var particles: [maxCapacity]Particle = undefined;
```